src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+<?php
+/**
+ * WordPress Plugin Administration API: WP_Rollback_Auto_Update class
+ *
+ * @package WordPress
+ * @subpackage Administration
+ * @since 6.3.0
+ */
+
+/**
+ * Core class for rolling back auto-update plugin failures.
+ */
+class WP_Rollback_Auto_Update {
+
+	/**
+	 * Stores handler parameters.
+	 *
+	 * @var array
+	 */
+	private $handler_args = array();
+
+	/**
+	 * Stores successfully updated plugins.
+	 *
+	 * @var array
+	 */
+	private static $processed = array();
+
+	/**
+	 * Stores fataling plugins.
+	 *
+	 * @var array
+	 */
+	private static $fatals = array();
+
+	/**
+	 * Stores `update_plugins` transient.
+	 *
+	 * @var stdClass
+	 */
+	private static $current;
+
+	/**
+	 * Stores boolean for no error from check_plugin_for_errors().
+	 *
+	 * @var bool
+	 */
+	private $no_error = false;
+
+	/**
+	 * Stores error codes.
+	 *
+	 * @var int
+	 */
+	public $error_types = E_ERROR | E_PARSE | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;
+
+	/**
+	 * Checks the validity of the updated plugin.
+	 *
+	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
+	 * @param array          $hook_extra Extra arguments passed to hooked filters.
+	 * @return array|WP_Error
+	 */
+	public function auto_update_check( $result, $hook_extra ) {
+		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
+			return $result;
+		}
+
+		// Already processed.
+		if ( in_array( $hook_extra['plugin'], self::$processed, true ) ) {
+			return $result;
+		}
+
+		error_log( $hook_extra['plugin'] . ' processing...' );
```

Comment:
Definitely worth having these logs for testing. Just a reminder for us to remove these before merge.

Reply:
Remove of put behind a WP_DEBUG is true?

Reply:
Don't think it needs to be shown with or without `WP_DEBUG` after merge, but for testing it might as well just get logged every time someone tests the PR